### ISSUE TITLE
Meta DEx trading interface

### ISSUE THREAD

**Issue by dexX7** [OmniLayer/omnicore#31]
![b34f1bc6-e432-11e4-85e5-9dbc6b443870](https://cloud.githubusercontent.com/assets/5836089/7464042/55c3dab4-f2bf-11e4-931b-aeddcebee44f.png)

What do you think should be added, removed, or changed?

To continue the discussion from the other thread:

> ... we shouldn't round the price in the order book, but we should show the effective unit price after a match is made.

I agree that showing the effective unit price is the way to go, and in general, I have no objections against a "very accurate" view of the available offers, but I'm wondering, if it's necessary, given that unit prices are shown "per MSC", and the input fields for the price are also "per MSC".

My first instinct was to show only up to 8 decimal places, as this aligns well with the minimal tradable amount of MSC, and then adjust the numbers such that a trade is executed, if the user creates an order with that price. For example: say there is an offer with an effective unit price of `0.10002555555555..`, then I'd assume, the user would enter either `0.10002555` or `0.10002556` to match that offer, and if this is the case, I was wondering, if we might just show the price rounded up or down to one willet right from the beginning?

**Comment by zathras-crypto** [comment]
Just as a heads up, I need to redo some of the UI data sources - the pricing is actually obtained from the price level in the metadex maps - but it turns out that was bad data because of the bug in persistence inverting the price.  We now have a ```.inversePrice()``` in the ```CMPMetaDEx``` class so I think I should just need to switch over to using that.

I'd like to get some kind of consolidation going at some point (especially Class C stuff)

**Comment by zathras-crypto** [comment]
> I was wondering, if we might just show the price rounded up or down to one willet right from the beginning?

I actually favour this methodology especially at the UI layer - there comes a point where ease of use trumps - look at DEx v1 - there is no volume because it's a PITA to use.  Thus I vote where things are 'safe' to do and improve the experience for the user (such as DexX's suggestion) then we go for it.

**Comment by zathras-crypto** [comment]
@dexX7 I'm just trying to improve the trading interface and noticed on another thread you mentioned:

> Swap buy and sell offers area (is there any exchange where this is the other way around?)
![image](https://cloud.githubusercontent.com/assets/5633351/7548621/760b06a6-f654-11e4-83c9-d8687fc06cfa.png)

That's an interesting point - I did actually poke around a few exchanges before designing it the way I did - it seemed logical that someone would be buying into a sell order, or selling into a buy order hence listing sell offers on the buy side and listing buy offers on the sell side.

Examples - BTC-e and Poloniex are like the way I originally designed it, Bitstamp is like your modified screenie above.

![image](https://cloud.githubusercontent.com/assets/5633351/7548623/afb42e28-f654-11e4-884c-b5ec101425a0.png)
![image](https://cloud.githubusercontent.com/assets/5633351/7548625/b483e98e-f654-11e4-80b4-ac41f576b07f.png)

Long story short I'd like to pick a direction - any further thoughts on this at all mate?

**Comment by dexX7** [comment]
> someone would be buying into a sell order, or selling into a buy order

Hmm.. this appears to be different from exchange to exchange. I think buy orders should be listed on the left, while sell orders should be listed on the right. This seems more natural (left to right, top to bottom). I'm unsure about the input masks.

**Comment by zathras-crypto** [comment]
Quick Q guys.  Do you think the UI needs to accommodate *all* potential scenarios on the MetaDEx?

Long story short I'm running into numerous complications due to the fact the "unit price" can be less that the minimum decimal fraction of an MSC.  So for example it is possible to create a trade with a unit price of:
```
0.000000000000000000000000001084 (0.00000001 tokens desired / 9,223,372,036,854,775,807 tokens for sale)
```
However since we have agreed to price everything in MSC, it is not valid to say allow the user to interact on a level that permits purchasing 1 unit per 0.000000000000000000000000001084 MSC, simply because the user cannot transact 0.000000000000000000000000001084 MSC.

What I guess I'm proposing is that (purely from a UI perspective) trades with a unit price under 0.00000001 MSC will not be tradeable via the UI.  They would not be shown, nor could they be acted upon or created (again UI only, you could still create such a trade at the RPC layer should you so desire).

This would permit the continuation of the model where all prices are shown in MSC, and not provide complexities for the user resulting from prices that are under that of the minimum transact-able amount of MSC (0.00000001 MSC).

I'm leaning in this direction because whilst the protocol does support these types of miniscule priced trades (eg 0.00000001 tokens desired / 9,223,372,036,854,775,807 tokens for sale) it is highly unlikely that an 'end user' at the UI layer is going to have a need to create or interact with them.

So, TL:DR; I'm considering making the UI limited to a preicision of 0.00000001 MSC, with prices rounded to the nearest willet and trades with prices under 1 willet considered 'advanced trades' and left out of the UI entirely.

What do you think?

**Comment by zathras-crypto** [comment]
The above doesn't seem to work particularly well either based on my local experiments :(

**Comment by zathras-crypto** [comment]
Note, I *really* don't want to abandon the "buy/sell" model we have for the UI at the moment because it's something users will be easily able to pick up - at the moment it's looking more and more like we need to ditch that completely and move to forcing the user to supply raw ```amountForSale``` and ```amountDesired``` values but I'm uber worried if we do this then MetaDEx will go the way of DEx v1 - ie no-one will use it because it's just so unfriendly and unintuitive :(

**Comment by zathras-crypto** [comment]
FYI, the problem with using rounding in the UI - we click a sell offer (which automatically sets up the buy):

![image](https://cloud.githubusercontent.com/assets/5633351/7604397/6aeb56a2-f985-11e4-8e05-4d58c0cb5000.png)

The price is actually ```0.10999999997051402461909080516077581355283271004378``` but comes to 0.11 when display precision is set to 8 decimals.  The UI must inherently setup trades based on the "price per SPT" and I fear issues like this will only cause user confusion.

**Comment by dexX7** [comment]
> The price is actually ...

Not all rational numbers are finite, so even this may not be precise enough.

I see the dilemma.

> I'm uber worried if we do this then MetaDEx will go the way of DEx v1.

It's hard to tell, how the core client is used, and my guess is that most users use Omni Wallet or other thin clients, and it's probably clear we can't solve all UI issues in the tiny QT interface.

I'd say: let's try to make the user experience as good as possible, even if this comes at the cost of limiting the user's actions (i.e. no trading of 0.0000001 for 2^63-1 or so), instead of trying to fit everything together, or falling back to "amount desired" and "amount for sale".

**Comment by patrickdugan** [comment]
Re: the very small numbers - I appreciate that this is a real math constraint but let's be realistic, how much of a use case is it for someone to buy a quantity valued at a millionth of a penny? (or less...)

Just clip it, disable it in the UI, solved. Omnicore logic-wise, you shouldn't be able to buy anything less than 1 willet worth of a smart property.

**Comment by zathras-crypto** [comment]
> the user needs to spend a touch over half a Mastercoin more than necessary

My apologies, this is not accurate - my experiments were based on the existing (```XDOUBLE```) MetaDEx logic, and we can see from the following that problematic math was responsible for such a big difference (I thought half a Mastercoin was *far, far* too big a skew just from enforcing 8 decimal pricing):

If we look at the trade (the one I circled in the screenie) directly in the map:
```0.10999999997051402461909080516077581355283271004378:trade #12 325.57851236 for #1 35.81363635```

And then look at our getter functions (also used by the UI):
```
displayUnitPrice() 0.11000000
displayInversePrice() 9.09090909
getAmountForSale() 32557851236
getAmountDesired() 3529363635
```

We can see that ```getAmountDesired()``` is ~0.5 MSC off what is in the map (```3529363635``` vs ```3581363635```).

TL:DR; looks like 8 decimal pricing might be workable after all in the UI without crazy skew (which I was worried would turn off users).  I'm about to start rewriting UI to use rationals as per @dexX7's excellent recent work and I'm hopeful this will be resolved.  Will update here when I have more :)

**Comment by dexX7** [comment]
Displaying only a certain number of decimal places, and restricting the number of decimal places to enter, goes hand in hand, as it seems.

Let's assume for a moment all users are UI users, and let's say all UI users can only enter prices with 8 decimal places, then all prices have 8 decimal places, right?

In practise it's unfortunally not that easy, because there are prices with more than 8 decimal places, which may create a disadvantage for UI users, and likely a rather odd impression, if `total MSC` doesn't equal `displayed price * total SPx`.

**Comment by zathras-crypto** [comment]
> Let's assume for a moment all users are UI users, and let's say all UI users can only enter prices with 8 decimal places, then all prices have 8 decimal places, right?

That's indeed correct :)

> In practise it's unfortunally not that easy, because there are prices with more than 8 decimal places

I'm looking into what kind of skew we're looking at here, and from my early experiments it should be a willet or less, but I need to do more to confirm this.

I'm currently looking at how to work with these new rational prices - specifically a couple of things:
1) converting them reliably to a string format with 8 decimal places (which I see you go over to dec_float first in ```xToString``` but only if the range is suitable for int64 so doesn't cover all cases)
2) the best approach to divide a rational by ```COIN``` - I see in the boost rational docs that the arithmetic operator ```/``` is valid so I may be able to do it directly

Early days for me while I poke around and learn a bit (I haven't worked with rationals like this before) but if you have any suggestions on 1 and 2 feel free to shout mate :)

> and likely a rather odd impression, if total MSC doesn't equal displayed price * total SPx.

This is exactly the behaviour I'm trying to avoid - again should know more soon :)

**Comment by dexX7** [comment]
The [rangeInt64()](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_mdex.h#L29-L37) checks were only relevant for intermediate results, in particular for the amount that one could buy, based on the seller's price and buyer's available amount, which can be larger than `int64_t`, but in the end all prices should be within [rangeInt64()](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_mdex.h#L29-L37).

> how to work with these new rational prices

In general, it's possible to do:

```c++
rational_t x(1, 2);   // 1 / 2
rational_t y = x / 3; // 1 / 2 / 3
rational_t z = price / COIN;
```

> converting them reliably to a string format with 8 decimal places

There are probably three options:

- [xToString()](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_mdex.h#L49-L55), which should return the price as string with 50 decimal places

- convert the price to `dec_float`, as done in [xToString()](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_mdex.h#L49-L55), and then work with the floating number

- or use [xToInt64()](https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_mdex.h#L75-L82) to convert to `int64_t` , which also rounds correctly in either way, and then format and convert to string via `FormatDivisibleMP()`, `FormatIndivisibleMP()` or `FormatMP()`

The last route is probably the easiest and most reliable.

**Comment by zathras-crypto** [comment]
Thanks for the advice mate :)  

I think I was getting myself confused (I'm a bit slow today!!!) eg ```trade #12 325.57851236 for #1 35.81363635``` is a price of ```~0.10999999997051402461909080516078``` and the fractional part of that number does not fit into ```int64_t``` - but I realize now that's irrelevant (it's the denominator and numerator for the rational that need to fit into ```int64_t```).

Thanks for walking me through it mate :)

**Comment by zathras-crypto** [comment]
> The last route is probably the easiest and most reliable.

No joy for now, passing the price into ```xToInt64``` makes every price = 1 willet - it's probably something straight forward but having a hard time focusing (bit distracted) - will tackle it again tomorrow with a set of fresh eyes :)
```
Unit Price: xToString=1.00000000000000000000000000000000000000000000000000 xToInt64=1 DividedByCoin=false
Unit Price: xToString=0.00000000333333333333333333333333333333333333333333 xToInt64=1 DividedByCoin=true
Unit Price: xToString=0.00000100000000000000000000000000000000000000000000 xToInt64=1 DividedByCoin=true
Unit Price: xToString=0.00000010000000000000000000000000000000000000000000 xToInt64=1 DividedByCoin=true
Unit Price: xToString=0.00000001000000000000000000000000000000000000000000 xToInt64=1 DividedByCoin=true
```

**Comment by dexX7** [comment]
> EDIT: scratch that, where the integer part of the prices is >1 ...

Ahh, wait. In this case `FormatDivisibleMP(xToInt64(price * COIN))` could work.

**Comment by zathras-crypto** [comment]
> Ahh, wait

It's not just about dividing by COIN mate, even those that we don't modify aren't coming out right - for example:
```
Display Unit Price Requested:
               Amount For Sale: 50000000 of 1
                Amount Desired: 5000000000 of 12
Unmodified Rational Unit Price: xToString=100.000000000000000000... xToInt64=100
    Divided by COIN Unit Price: xToString=100.000000000000000000... xToInt64=100 DividedByCoin=false
Unmodified Rational Invs Price: xToString=0.01000000000000000000... xToInt64=1
    Divided by COIN Invs Price: xToString=0.01000000000000000000... xToInt64=1 DividedByCoin=false
```
Looking into it a bit further now - haven't spent the day having tests today so hopefully I'll be in a clearer state of mind hehe!!! :)

**Comment by zathras-crypto** [comment]
OK, think I have it licked with https://github.com/zathras-crypto/mastercore/commit/cb57e5a1a3af96eb48de3be70fa3512c6f0fc244

The good news is after @dexX7's fixes if we now revisit that same trade we can see that unit price rounding to 8 decimal places doesn't seem to pose an issue for the calculated total.  There may be some potential cases where we may see up to a 1 willet skew but I don't think that'll scare off users.

![image](https://cloud.githubusercontent.com/assets/5633351/7646114/23e8d83c-fb02-11e4-933f-81915c8a6631.png)

**Comment by dexX7** [comment]
I'm very happy that it works. :)

> There may be some potential cases where we may see up to a 1 willet skew but I don't think that'll scare off users.

Could you document one of such cases, if you stumble upon one? I agree, one willet isn't a big deal, if the displayed amount is rounded such that an order fills nevertheless. What I'd like to avoid is a situation where an user buys/sells tokens at the displayed price, but then misses by one willet.

**Comment by dexX7** [comment]
I currently see no specifics to discuss, and once there is, we can open a new issue. /close
